trlx/models/modeling_nemo_ppo.py

```diff
@@ -191,93 +169,114 @@ def __init__(self, language_model, other_heads, build_reference_model=True):
 
     def word_embeddings_weight(self):
         return self._lm.word_embeddings_weight()
 
     def load_state_dict(self, lm_state_dict, strict=True):
         """Load GPTModel state dict."""
         self.language_model.load_state_dict(lm_state_dict, strict=strict)
-        if self.reference_model is not None:
-            self.reference_model.model.language_model.load_state_dict(lm_state_dict, strict=strict)
+
+        if self.build_reference_model:
+            for p in self.language_model.parameters():
+                if p.requires_grad:
+                    p._ref_data = p.data.clone().cpu().pin_memory()
+                    p._policy_data = p.data
 
     def pretrained_state_dict(self):
         """Load GPTModel state dict."""
         return self._lm.state_dict()
 
     def offload_reference_model(self):
         """Move reference model to CPU."""
-        self.reference_model.offload()
-        self._lm.to(torch.cuda.current_device(), non_blocking=True)
+        if self.reference_model_offloaded or not self.build_reference_model:
+            return
+        for p in self.language_model.parameters():
+            if p.requires_grad:
+                p.data = p._policy_data.to(torch.cuda.current_device())
+        self.reference_model_offloaded = True
 
     def offload_policy_model(self):
         """Move language model to CPU."""
-        self.reference_model.onload()
-        self._lm.to("cpu", non_blocking=True)
+        if self.reference_model_offloaded:
+            for p in self.language_model.parameters():
+                if p.requires_grad:
+                    p._policy_data = p.data.to("cpu", non_blocking=True)
+                    p.data = p._ref_data.to(torch.cuda.current_device(), non_blocking=True)
+            self.reference_model_offloaded = False
 
     def forward(
         self,
         *args,
         get_key_value=False,
         forward_method_parallel_output=None,
+        run_policy_model=True,
         run_reference_model=False,
         run_value_head=False,
         **kwargs,
     ):
-        lm_output = self._lm(*args, get_key_value=get_key_value, **kwargs)
-        logits = post_language_model_processing(
-            lm_output,
-            labels=None,
-            logit_weights=self._lm.word_embeddings_weight(),
-            get_key_value=get_key_value,
-            parallel_output=False,  # self.language_model.parallel_output,
-            forward_method_parallel_output=forward_method_parallel_output,
-            fp16_lm_cross_entropy=self._lm.fp16_lm_cross_entropy,
-            return_logits=True,
-            sequence_parallel=self._lm.sequence_parallel,
-            gradient_accumulation_fusion=self._lm.gradient_accumulation_fusion,
-        )
+        logit_weights = self._lm.word_embeddings_weight()
 
-        if get_key_value:
-            logits, presents = logits
-            lm_output, lm_output_presents = lm_output
+        if run_policy_model:
+            self.offload_reference_model()
```

Comment:
Does the reference model get offloaded more than once per execution?

Reply:
Nevermind I see

Reply:
Ah it should be memozied offloading